File: .claude/examples/python/internal_tables_migration.py
"""
ABAP Internal Tables to Python Collections Migration Example

This module demonstrates how to migrate common ABAP internal table operations
to equivalent Python code using modern Python idioms.

Based on: src/zcl_demo_abap_internal_tables.clas.abap
Reference: 01_Internal_Tables.md
"""

from dataclasses import dataclass, field
from typing import List, Optional, Dict, Any
from enum import Enum
from datetime import date
import itertools


# ==============================================================================
# Data Models (ABAP Types → Python dataclasses)
# ==============================================================================

@dataclass
class Flight:
    """
    Equivalent to ABAP structure for flight data
    ABAP: TYPES ty_flight TYPE zdemo_abap_fli
    """
    carrid: str  # Carrier ID
    connid: str  # Connection ID
    fldate: date  # Flight date
    price: float  # Price
    currency: str  # Currency
    seatsmax: int  # Maximum seats
    seatsocc: int  # Occupied seats


@dataclass
class Carrier:
    """
    Equivalent to ABAP carrier structure
    ABAP: TYPES ty_carrier TYPE zdemo_abap_carr
    """
    carrid: str  # Carrier ID
    carrname: str  # Carrier name
    currcode: str  # Currency code
    url: str = ""  # Website URL


# ==============================================================================
# Pattern 1: ABAP APPEND → Python list.append()
# ==============================================================================

def example_append_operation():
    """
    ABAP:
        DATA lt_flights TYPE TABLE OF ty_flight.
        APPEND VALUE #( carrid = 'AA' connid = '001' ) TO lt_flights.

    Python equivalent:
    """
    flights: List[Flight] = []

    # Simple append
    flights.append(Flight(
        carrid='AA',
        connid='001',
        fldate=date(2024, 1, 15),
        price=450.00,
        currency='USD',
        seatsmax=200,
        seatsocc=150
    ))

    return flights


# ==============================================================================
# Pattern 2: ABAP LOOP AT → Python for loop
# ==============================================================================

def example_loop_operations(flights: List[Flight]) -> List[Flight]:
    """
    ABAP:
        LOOP AT lt_flights INTO DATA(ls_flight).
          IF ls_flight-seatsmax > 100.
            ls_flight-price = ls_flight-price * '1.1'.
            MODIFY lt_flights FROM ls_flight.
          ENDIF.
        ENDLOOP.

    Python equivalent:
    """
    # Method 1: Direct modification (flights are mutable objects)
    for flight in flights:
        if flight.seatsmax > 100:
            flight.price = flight.price * 1.1

    # Method 2: Functional approach (immutable)
    updated_flights = [
        Flight(
            **{**flight.__dict__, 'price': flight.price * 1.1}
        ) if flight.seatsmax > 100 else flight
        for flight in flights
    ]

    return flights


# ==============================================================================
# Pattern 3: ABAP READ TABLE → Python filtering
# ==============================================================================

def example_read_table(flights: List[Flight], carrier: str) -> Optional[Flight]:
    """
    ABAP:
        READ TABLE lt_flights INTO DATA(ls_flight)
          WITH KEY carrid = 'AA'.

    Python equivalent:
    """
    # Method 1: Using next() with default
    flight = next(
        (f for f in flights if f.carrid == carrier),
        None
    )

    # Method 2: Using filter (if you need the first match)
    try:
        flight = next(filter(lambda f: f.carrid == carrier, flights))
    except StopIteration:
        flight = None

    return flight


# ==============================================================================
# Pattern 4: ABAP FILTER → Python list comprehension / filter()
# ==============================================================================

def example_filter_table(flights: List[Flight], min_seats: int = 150) -> List[Flight]:
    """
    ABAP:
        DATA(lt_filtered) = FILTER #( lt_flights
          WHERE seatsmax > 150 ).

    Python equivalent:
    """
    # Method 1: List comprehension (Pythonic)
    filtered = [f for f in flights if f.seatsmax > min_seats]

    # Method 2: filter() function
    filtered_alt = list(filter(lambda f: f.seatsmax > min_seats, flights))

    return filtered


# ==============================================================================
# Pattern 5: ABAP SORT → Python sorted() / list.sort()
# ==============================================================================

def example_sort_table(flights: List[Flight]) -> List[Flight]:
    """
    ABAP:
        SORT lt_flights BY carrid ASCENDING connid DESCENDING.

    Python equivalent:
    """
    # Method 1: In-place sort
    flights.sort(key=lambda f: (f.carrid, -float(f.connid)))

    # Method 2: Create new sorted list
    sorted_flights = sorted(
        flights,
        key=lambda f: (f.carrid, f.connid),
        reverse=False
    )

    # Method 3: Multiple sort keys (more complex)
    from operator import attrgetter
    sorted_flights = sorted(
        flights,
        key=lambda f: (f.carrid, f.connid)
    )

    return sorted_flights


# ==============================================================================
# Pattern 6: ABAP DELETE → Python list comprehension / filter()
# ==============================================================================

def example_delete_from_table(flights: List[Flight], carrier_to_remove: str) -> List[Flight]:
    """
    ABAP:
        DELETE lt_flights WHERE carrid = 'AA'.

    Python equivalent:
    """
    # Method 1: List comprehension (creates new list)
    flights = [f for f in flights if f.carrid != carrier_to_remove]

    # Method 2: In-place removal (less efficient for multiple items)
    flights_copy = flights.copy()
    for flight in flights_copy[:]:
        if flight.carrid == carrier_to_remove:
            flights_copy.remove(flight)

    return flights


# ==============================================================================
# Pattern 7: ABAP GROUP BY → Python itertools.groupby
# ==============================================================================

def example_group_by(flights: List[Flight]) -> Dict[str, List[Flight]]:
    """
    ABAP:
        DATA(lt_grouped) = VALUE tt_grouped(
          FOR GROUPS <group> OF <flight> IN lt_flights
          GROUP BY ( carrid = <flight>-carrid )
          ( carrid = <group>-carrid
            flights = VALUE #( FOR f IN GROUP <group> ( f ) ) )
        ).

    Python equivalent:
    """
    # Method 1: Using itertools.groupby (requires sorted data)
    flights_sorted = sorted(flights, key=lambda f: f.carrid)
    grouped = {
        carrier: list(group)
        for carrier, group in itertools.groupby(flights_sorted, key=lambda f: f.carrid)
    }

    # Method 2: Using dict comprehension
    carriers = set(f.carrid for f in flights)
    grouped_alt = {
        carrier: [f for f in flights if f.carrid == carrier]
        for carrier in carriers
    }

    return grouped


# ==============================================================================
# Pattern 8: ABAP REDUCE → Python reduce() / comprehensions
# ==============================================================================

def example_reduce_aggregation(flights: List[Flight]) -> float:
    """
    ABAP:
        DATA(lv_total) = REDUCE i(
          INIT sum = 0
          FOR flight IN lt_flights
          NEXT sum = sum + flight-price ).

    Python equivalent:
    """
    # Method 1: Using sum() (most Pythonic)
    total = sum(f.price for f in flights)

    # Method 2: Using reduce()
    from functools import reduce
    total_alt = reduce(lambda acc, f: acc + f.price, flights, 0)

    return total


# ==============================================================================
# Pattern 9: ABAP VALUE Constructor → Python list/dict comprehension
# ==============================================================================

def example_value_constructor(carrier_ids: List[str]) -> List[Flight]:
    """
    ABAP:
        DATA(lt_flights) = VALUE tt_flights(
          FOR carrier IN lt_carrier_ids
          ( carrid = carrier
            connid = '001'
            seatsmax = 200 )
        ).

    Python equivalent:
    """
    flights = [
        Flight(
            carrid=carrier,
            connid='001',
            fldate=date(2024, 1, 1),
            price=500.0,
            currency='USD',
            seatsmax=200,
            seatsocc=0
        )
        for carrier in carrier_ids
    ]

    return flights


# ==============================================================================
# Pattern 10: ABAP CORRESPONDING → Python dict unpacking
# ==============================================================================

@dataclass
class FlightSummary:
    """Simplified flight structure"""
    carrid: str
    connid: str
    price: float


def example_corresponding(flights: List[Flight]) -> List[FlightSummary]:
    """
    ABAP:
        DATA(lt_summary) = CORRESPONDING tt_summary(
          lt_flights MAPPING carrid = carrid connid = connid ).

    Python equivalent:
    """
    # Method 1: Manual field selection
    summaries = [
        FlightSummary(
            carrid=f.carrid,
            connid=f.connid,
            price=f.price
        )
        for f in flights
    ]

    # Method 2: Dynamic with shared fields
    summary_fields = set(FlightSummary.__dataclass_fields__.keys())
    summaries_alt = [
        FlightSummary(**{
            k: v for k, v in f.__dict__.items()
            if k in summary_fields
        })
        for f in flights
    ]

    return summaries


# ==============================================================================
# Complete Example: Complex Table Operations
# ==============================================================================

def complex_table_operations_example():
    """
    Demonstrates complex ABAP table operations translated to Python
    Mimics patterns from zcl_demo_abap_internal_tables
    """
    # Initialize sample data
    flights = [
        Flight('AA', '001', date(2024, 1, 15), 450.0, 'USD', 200, 150),
        Flight('AA', '002', date(2024, 1, 16), 480.0, 'USD', 180, 170),
        Flight('LH', '001', date(2024, 1, 15), 520.0, 'EUR', 220, 200),
        Flight('BA', '001', date(2024, 1, 17), 390.0, 'GBP', 150, 120),
    ]

    # Filter high-capacity flights
    large_flights = [f for f in flights if f.seatsmax >= 180]

    # Sort by carrier and connection
    large_flights.sort(key=lambda f: (f.carrid, f.connid))

    # Group by carrier and calculate totals
    flights_by_carrier = {}
    for flight in large_flights:
        if flight.carrid not in flights_by_carrier:
            flights_by_carrier[flight.carrid] = []
        flights_by_carrier[flight.carrid].append(flight)

    # Calculate revenue per carrier
    revenue_by_carrier = {
        carrier: sum(f.price * f.seatsocc for f in carrier_flights)
        for carrier, carrier_flights in flights_by_carrier.items()
    }

    print(f"Total carriers: {len(revenue_by_carrier)}")
    for carrier, revenue in revenue_by_carrier.items():
        print(f"{carrier}: ${revenue:,.2f}")

    return revenue_by_carrier


# ==============================================================================
# Performance Considerations
# ==============================================================================

def performance_tips():
    """
    Performance equivalents for ABAP table operations

    1. ABAP Sorted Tables → Python: Use bisect module or maintain sorted order
    2. ABAP Hashed Tables → Python: Use dict or set for O(1) lookups
    3. ABAP BINARY SEARCH → Python: Use bisect.bisect_left()
    4. Large datasets → Consider pandas DataFrame
    """
    import bisect

    # Example: Maintaining sorted table with binary search
    flights_sorted = []

    def insert_sorted(flight: Flight):
        """Insert maintaining sort order (like ABAP sorted table)"""
        bisect.insort(flights_sorted, flight, key=lambda f: f.carrid)

    # Example: Hash table for fast lookups (like ABAP hashed table)
    flights_dict: Dict[tuple, Flight] = {}

    def add_to_hash(flight: Flight):
        """Add to hash table with composite key"""
        key = (flight.carrid, flight.connid, flight.fldate)
        flights_dict[key] = flight

    def lookup_by_key(carrid: str, connid: str, fldate: date) -> Optional[Flight]:
        """O(1) lookup like ABAP hashed table READ"""
        return flights_dict.get((carrid, connid, fldate))


if __name__ == '__main__':
    print("ABAP Internal Tables to Python Migration Examples")
    print("=" * 60)
    complex_table_operations_example()

File: .claude/examples/typescript/internal-tables-migration.ts
/**
 * ABAP Internal Tables to TypeScript Collections Migration Example
 *
 * This module demonstrates how to migrate common ABAP internal table operations
 * to equivalent TypeScript code using modern JavaScript/TypeScript idioms.
 *
 * Based on: src/zcl_demo_abap_internal_tables.clas.abap
 * Reference: 01_Internal_Tables.md
 */

// ==============================================================================
// Data Models (ABAP Types → TypeScript interfaces/classes)
// ==============================================================================

/**
 * Equivalent to ABAP structure for flight data
 * ABAP: TYPES ty_flight TYPE zdemo_abap_fli
 */
interface Flight {
    carrid: string;      // Carrier ID
    connid: string;      // Connection ID
    fldate: Date;        // Flight date
    price: number;       // Price
    currency: string;    // Currency
    seatsmax: number;    // Maximum seats
    seatsocc: number;    // Occupied seats
}

/**
 * Equivalent to ABAP carrier structure
 * ABAP: TYPES ty_carrier TYPE zdemo_abap_carr
 */
interface Carrier {
    carrid: string;      // Carrier ID
    carrname: string;    // Carrier name
    currcode: string;    // Currency code
    url?: string;        // Website URL
}

// ==============================================================================
// Pattern 1: ABAP APPEND → TypeScript array.push()
// ==============================================================================

function exampleAppendOperation(): Flight[] {
    /**
     * ABAP:
     *   DATA lt_flights TYPE TABLE OF ty_flight.
     *   APPEND VALUE #( carrid = 'AA' connid = '001' ) TO lt_flights.
     *
     * TypeScript equivalent:
     */
    const flights: Flight[] = [];

    // Simple push
    flights.push({
        carrid: 'AA',
        connid: '001',
        fldate: new Date('2024-01-15'),
        price: 450.00,
        currency: 'USD',
        seatsmax: 200,
        seatsocc: 150
    });

    return flights;
}

// ==============================================================================
// Pattern 2: ABAP LOOP AT → TypeScript for...of / forEach
// ==============================================================================

function exampleLoopOperations(flights: Flight[]): Flight[] {
    /**
     * ABAP:
     *   LOOP AT lt_flights INTO DATA(ls_flight).
     *     IF ls_flight-seatsmax > 100.
     *       ls_flight-price = ls_flight-price * '1.1'.
     *       MODIFY lt_flights FROM ls_flight.
     *     ENDIF.
     *   ENDLOOP.
     *
     * TypeScript equivalent:
     */

    // Method 1: Direct modification (mutable)
    for (const flight of flights) {
        if (flight.seatsmax > 100) {
            flight.price = flight.price * 1.1;
        }
    }

    // Method 2: Functional approach (immutable)
    const updatedFlights = flights.map(flight =>
        flight.seatsmax > 100
            ? { ...flight, price: flight.price * 1.1 }
            : flight
    );

    return flights;
}

// ==============================================================================
// Pattern 3: ABAP READ TABLE → TypeScript find()
// ==============================================================================

function exampleReadTable(flights: Flight[], carrier: string): Flight | undefined {
    /**
     * ABAP:
     *   READ TABLE lt_flights INTO DATA(ls_flight)
     *     WITH KEY carrid = 'AA'.
     *
     * TypeScript equivalent:
     */

    // Method 1: Using find()
    const flight = flights.find(f => f.carrid === carrier);

    // Method 2: Using filter()[0]
    const flightAlt = flights.filter(f => f.carrid === carrier)[0];

    return flight;
}

// ==============================================================================
// Pattern 4: ABAP FILTER → TypeScript filter()
// ==============================================================================

function exampleFilterTable(flights: Flight[], minSeats: number = 150): Flight[] {
    /**
     * ABAP:
     *   DATA(lt_filtered) = FILTER #( lt_flights
     *     WHERE seatsmax > 150 ).
     *
     * TypeScript equivalent:
     */

    // Method 1: Using filter()
    const filtered = flights.filter(f => f.seatsmax > minSeats);

    // Method 2: With multiple conditions
    const filteredComplex = flights.filter(f =>
        f.seatsmax > minSeats &&
        f.seatsocc < f.seatsmax * 0.9
    );

    return filtered;
}

// ==============================================================================
// Pattern 5: ABAP SORT → TypeScript sort()
// ==============================================================================

function exampleSortTable(flights: Flight[]): Flight[] {
    /**
     * ABAP:
     *   SORT lt_flights BY carrid ASCENDING connid DESCENDING.
     *
     * TypeScript equivalent:
     */

    // Method 1: In-place sort
    flights.sort((a, b) => {
        // Sort by carrid ascending
        if (a.carrid < b.carrid) return -1;
        if (a.carrid > b.carrid) return 1;

        // Then by connid descending
        if (a.connid > b.connid) return -1;
        if (a.connid < b.connid) return 1;

        return 0;
    });

    // Method 2: Create new sorted array
    const sortedFlights = [...flights].sort((a, b) =>
        a.carrid.localeCompare(b.carrid) || b.connid.localeCompare(a.connid)
    );

    return sortedFlights;
}

// ==============================================================================
// Pattern 6: ABAP DELETE → TypeScript filter()
// ==============================================================================

function exampleDeleteFromTable(flights: Flight[], carrierToRemove: string): Flight[] {
    /**
     * ABAP:
     *   DELETE lt_flights WHERE carrid = 'AA'.
     *
     * TypeScript equivalent:
     */

    // Method 1: Using filter (creates new array)
    const remainingFlights = flights.filter(f => f.carrid !== carrierToRemove);

    // Method 2: In-place removal using splice (less efficient)
    for (let i = flights.length - 1; i >= 0; i--) {
        if (flights[i].carrid === carrierToRemove) {
            flights.splice(i, 1);
        }
    }

    return remainingFlights;
}

// ==============================================================================
// Pattern 7: ABAP GROUP BY → TypeScript reduce() / Map
// ==============================================================================

function exampleGroupBy(flights: Flight[]): Map<string, Flight[]> {
    /**
     * ABAP:
     *   DATA(lt_grouped) = VALUE tt_grouped(
     *     FOR GROUPS <group> OF <flight> IN lt_flights
     *     GROUP BY ( carrid = <flight>-carrid )
     *     ( carrid = <group>-carrid
     *       flights = VALUE #( FOR f IN GROUP <group> ( f ) ) )
     *   ).
     *
     * TypeScript equivalent:
     */

    // Method 1: Using reduce()
    const grouped = flights.reduce((acc, flight) => {
        const carrier = flight.carrid;
        if (!acc.has(carrier)) {
            acc.set(carrier, []);
        }
        acc.get(carrier)!.push(flight);
        return acc;
    }, new Map<string, Flight[]>());

    // Method 2: Using for...of
    const groupedAlt = new Map<string, Flight[]>();
    for (const flight of flights) {
        const carrier = flight.carrid;
        if (!groupedAlt.has(carrier)) {
            groupedAlt.set(carrier, []);
        }
        groupedAlt.get(carrier)!.push(flight);
    }

    // Method 3: Using lodash (if available)
    // import { groupBy } from 'lodash';
    // const groupedLodash = groupBy(flights, 'carrid');

    return grouped;
}

// ==============================================================================
// Pattern 8: ABAP REDUCE → TypeScript reduce()
// ==============================================================================

function exampleReduceAggregation(flights: Flight[]): number {
    /**
     * ABAP:
     *   DATA(lv_total) = REDUCE i(
     *     INIT sum = 0
     *     FOR flight IN lt_flights
     *     NEXT sum = sum + flight-price ).
     *
     * TypeScript equivalent:
     */

    // Method 1: Using reduce()
    const total = flights.reduce((sum, flight) => sum + flight.price, 0);

    // Method 2: Using for...of
    let totalAlt = 0;
    for (const flight of flights) {
        totalAlt += flight.price;
    }

    return total;
}

// ==============================================================================
// Pattern 9: ABAP VALUE Constructor → TypeScript map()
// ==============================================================================

function exampleValueConstructor(carrierIds: string[]): Flight[] {
    /**
     * ABAP:
     *   DATA(lt_flights) = VALUE tt_flights(
     *     FOR carrier IN lt_carrier_ids
     *     ( carrid = carrier
     *       connid = '001'
     *       seatsmax = 200 )
     *   ).
     *
     * TypeScript equivalent:
     */

    const flights = carrierIds.map(carrier => ({
        carrid: carrier,
        connid: '001',
        fldate: new Date('2024-01-01'),
        price: 500.0,
        currency: 'USD',
        seatsmax: 200,
        seatsocc: 0
    }));

    return flights;
}

// ==============================================================================
// Pattern 10: ABAP CORRESPONDING → TypeScript destructuring/mapping
// ==============================================================================

interface FlightSummary {
    carrid: string;
    connid: string;
    price: number;
}

function exampleCorresponding(flights: Flight[]): FlightSummary[] {
    /**
     * ABAP:
     *   DATA(lt_summary) = CORRESPONDING tt_summary(
     *     lt_flights MAPPING carrid = carrid connid = connid ).
     *
     * TypeScript equivalent:
     */

    // Method 1: Manual field selection
    const summaries = flights.map(f => ({
        carrid: f.carrid,
        connid: f.connid,
        price: f.price
    }));

    // Method 2: Using destructuring
    const summariesAlt = flights.map(({ carrid, connid, price }) => ({
        carrid,
        connid,
        price
    }));

    return summaries;
}

// ==============================================================================
// Complete Example: Complex Table Operations
// ==============================================================================

function complexTableOperationsExample(): Map<string, number> {
    /**
     * Demonstrates complex ABAP table operations translated to TypeScript
     * Mimics patterns from zcl_demo_abap_internal_tables
     */

    // Initialize sample data
    const flights: Flight[] = [
        { carrid: 'AA', connid: '001', fldate: new Date('2024-01-15'), price: 450.0, currency: 'USD', seatsmax: 200, seatsocc: 150 },
        { carrid: 'AA', connid: '002', fldate: new Date('2024-01-16'), price: 480.0, currency: 'USD', seatsmax: 180, seatsocc: 170 },
        { carrid: 'LH', connid: '001', fldate: new Date('2024-01-15'), price: 520.0, currency: 'EUR', seatsmax: 220, seatsocc: 200 },
        { carrid: 'BA', connid: '001', fldate: new Date('2024-01-17'), price: 390.0, currency: 'GBP', seatsmax: 150, seatsocc: 120 },
    ];

    // Filter high-capacity flights
    const largeFlights = flights.filter(f => f.seatsmax >= 180);

    // Sort by carrier and connection
    largeFlights.sort((a, b) =>
        a.carrid.localeCompare(b.carrid) || a.connid.localeCompare(b.connid)
    );

    // Group by carrier
    const flightsByCarrier = largeFlights.reduce((acc, flight) => {
        if (!acc.has(flight.carrid)) {
            acc.set(flight.carrid, []);
        }
        acc.get(flight.carrid)!.push(flight);
        return acc;
    }, new Map<string, Flight[]>());

    // Calculate revenue per carrier
    const revenueByCarrier = new Map<string, number>();
    for (const [carrier, carrierFlights] of flightsByCarrier.entries()) {
        const revenue = carrierFlights.reduce(
            (sum, f) => sum + (f.price * f.seatsocc),
            0
        );
        revenueByCarrier.set(carrier, revenue);
    }

    console.log(`Total carriers: ${revenueByCarrier.size}`);
    for (const [carrier, revenue] of revenueByCarrier.entries()) {
        console.log(`${carrier}: $${revenue.toLocaleString('en-US', { minimumFractionDigits: 2 })}`);
    }

    return revenueByCarrier;
}

// ==============================================================================
// Performance Considerations
// ==============================================================================

/**
 * Performance equivalents for ABAP table operations
 *
 * 1. ABAP Sorted Tables → TypeScript: Maintain sorted array or use binary search
 * 2. ABAP Hashed Tables → TypeScript: Use Map or Set for O(1) lookups
 * 3. ABAP BINARY SEARCH → TypeScript: Implement binary search or use libraries
 * 4. Large datasets → Consider using specialized libraries like lodash
 */

class SortedFlightTable {
    private flights: Flight[] = [];

    /**
     * Insert maintaining sort order (like ABAP sorted table)
     */
    insertSorted(flight: Flight): void {
        const index = this.flights.findIndex(f => f.carrid > flight.carrid);
        if (index === -1) {
            this.flights.push(flight);
        } else {
            this.flights.splice(index, 0, flight);
        }
    }

    /**
     * Binary search for flights by carrier
     */
    findByCarrier(carrid: string): Flight | undefined {
        let left = 0;
        let right = this.flights.length - 1;

        while (left <= right) {
            const mid = Math.floor((left + right) / 2);
            const midCarrier = this.flights[mid].carrid;

            if (midCarrier === carrid) {
                return this.flights[mid];
            } else if (midCarrier < carrid) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return undefined;
    }
}

/**
 * Hash table for fast lookups (like ABAP hashed table)
 */
class HashedFlightTable {
    private flightsMap = new Map<string, Flight>();

    /**
     * Add to hash table with composite key
     */
    add(flight: Flight): void {
        const key = `${flight.carrid}|${flight.connid}|${flight.fldate.toISOString()}`;
        this.flightsMap.set(key, flight);
    }

    /**
     * O(1) lookup like ABAP hashed table READ
     */
    lookupByKey(carrid: string, connid: string, fldate: Date): Flight | undefined {
        const key = `${carrid}|${connid}|${fldate.toISOString()}`;
        return this.flightsMap.get(key);
    }
}

// ==============================================================================
// Main Execution
// ==============================================================================

if (require.main === module) {
    console.log('ABAP Internal Tables to TypeScript Migration Examples');
    console.log('='.repeat(60));
    complexTableOperationsExample();
}

export {
    Flight,
    Carrier,
    FlightSummary,
    exampleAppendOperation,
    exampleLoopOperations,
    exampleReadTable,
    exampleFilterTable,
    exampleSortTable,
    exampleDeleteFromTable,
    exampleGroupBy,
    exampleReduceAggregation,
    exampleValueConstructor,
    exampleCorresponding,
    complexTableOperationsExample,
    SortedFlightTable,
    HashedFlightTable
};
